igcommit/file_checks.py

```diff
@@ -104,14 +107,37 @@ def get_exe_problems(self, extension):
                     .format(exe, key)
                 )
 
-    def __str__(self):
-        return '{} on {}'.format(type(self).__name__, self.committed_file)
 
+class CommittedFileByExtensionCheck(CommittedFileCheck):
+    extension = None
 
-class CheckCommand(CommmittedFileCheck):
+    def prepare(self, obj):
+        new = super(CommittedFileByExtensionCheck, self).prepare(obj)
+        if not new or not isinstance(obj, CommittedFile):
+            return new
+
+        # All instances of this must specify a file extension.
+        assert new.extension
+
+        # We are being prepared for a committed file as last.  In this step,
+        # we need to match the file with the specified file extension.
+        # We first check the extension from the name of the file, and then
+        # from the shebang of the file, if it is executable.
+        if obj.get_extension() == new.extension:
+            return new
+        if (
+            new.extension in FILE_EXTENSIONS and
+            obj.owner_can_execute() and
+            FILE_EXTENSIONS[new.extension].search(obj.get_exe())
+        ):
+            return new
```

Comment:
Way better, especially with the comments. However I would prefer the comments directly at the if statements if possible. I also prefer `None-Checks` so `None` will be returned in the non-normal behavior with the `if-statements`. It would more complicate the second statement again though, so I am not sure right now.
It is fair enough that way generally speaking.